packages/overlay/src/Overlay.ts

```diff
@@ -0,0 +1,156 @@
+};
+
+/**
+ * @element sp-overlay
+ *
+ * @fires sp-opened - announces that an overlay has completed any entry animations
+ * @fires sp-closed - announce that an overlay has compelted any exit animations
+ */
+export class Overlay extends OverlayFeatures implements OverlayBase {
+    public static update(): void {
+        const overlayUpdateEvent = new CustomEvent('sp-update-overlays', {
+            bubbles: true,
+            composed: true,
+            cancelable: true,
+        });
+        document.dispatchEvent(overlayUpdateEvent);
+    }
+
+    public static async open(
+        target: HTMLElement,
+        interaction: TriggerInteractionsV1,
+        content: HTMLElement,
+        options: OverlayOptionsV1
+    ): Promise<() => void>;
+    public static async open(
+        content: HTMLElement,
+        options?: OverlayOptions
+    ): Promise<Overlay>;
+    public static async open(
+        targetOrContent: HTMLElement,
+        interactionOrOptions:
+            | TriggerInteractionsV1
+            | OverlayOptions
+            | undefined,
+        content?: HTMLElement,
+        options?: OverlayOptionsV1
+    ): Promise<Overlay | (() => void)> {
+        const v2 = arguments.length === 2;
+        const overlayContent = content || targetOrContent;
+        const overlay = new Overlay();
+        overlay.dispose = () => {
+            overlay.addEventListener('sp-closed', () => {
+                if (!restored) {
+                    restoreContent();
+                    restored = true;
+                }
+                requestAnimationFrame(() => {
+                    overlay.remove();
+                });
+            });
+            overlay.open = false;
+            overlay.dispose = noop;
+        };
+        let restored = false;
+        const restoreContent = reparentChildren([overlayContent], overlay, {
+            position: 'beforeend',
+            prepareCallback: (el) => {
+                const slot = el.slot;
+                el.removeAttribute('slot');
+                return () => {
+                    el.slot = slot;
+                };
+            },
+        });
+        if (v2) {
+            const options = interactionOrOptions as OverlayOptions;
+            overlay.append(overlayContent);
+            overlay.receivesFocus = options.receivesFocus ?? 'auto';
+            overlay.triggerElement = options.trigger || null;
+            overlay.type = options.type || 'modal';
+            overlay.offset = options.offset || 6;
+            overlay.placement = options.placement;
+            overlay.willPreventClose = !!options.notImmediatelyClosable;
+            requestAnimationFrame(() => {
+                requestAnimationFrame(() => {
+                    // Do we want to "open" this path, or leave that to the consumer?
+                    overlay.open = true;
+                });
+            });
+            return overlay;
+        } else if (overlayContent && options) {
```

Comment:
Do we expect this to run anytime we're exiting early via "v2" mode? If so, maybe we should delete the branch & satisfy TS that the required values exist like:
 
```suggestion
const v1 = !v2 && overlayContent && options;
if (!v1) {
  // good place for a dev mode warning about satisfying neither V1 nor V2 overlay argument requirements
  return;
}
```

Reply:
Edit: I see sometimes we `return overlay` and sometimes `return overlay.dispose`, so maybe this *is* an optional branch and not just the v1 fallback.